src/exp2.jl

```diff
@@ -0,0 +1,66 @@
+# Import the generated code
+for i=1:15
+    include("exp_generated/exp_$i.jl")
+end
+
+function getmem(cache,k)
+    return cache[k-1];
+end
+
+function _exp2!(A; caches=nothing, do_balancing = A isa StridedMatrix)
+    n = LinearAlgebra.checksquare(A)
+    nA = opnorm(A,1);
+
+    # Maybe to balancing
+    if do_balancing
+        ilo, ihi, scale = LAPACK.gebal!('B', A)    # modifies A
+    end
+
+    # Select how many multiplications to use
+    rhov=    [0.015; 0.25; 0.95; 2.1; 5.4];
+    # Number of memslots needed (beside A)
+    memslots=[3   ;    4;    5;   6;   5];
+    for s=1:8
+        push!(rhov, rhov[end]*2);
+        push!(memslots,memslots[end]);
+    end
+    i = findfirst(nA .< rhov)
+
+    # Take care of caching
+    if caches == nothing
+        # No cache. Allocate Matrix objects
+        TT=eltype(A);
+        thiscache=[Matrix{TT}(undef,n,n) for i=1:memslots[i]];
+    else
+        thiscache = caches;
+    end
+
+
+    # Evaluate
+    exp_fun=Symbol("exp_$(i)!")
+    s=:($(exp_fun)($thiscache,$A));
+    X=eval(s)::typeof(A) # Evaluate and maintain type stability
+
```

Comment:
no need for eval here, just branch.

Reply:
I don't understand. What do you mean by branch?

Reply:
`if` statements. Relying on `eval` is just a very indirect if statement that will have more overhead

Reply:
You mean like this? 
```
if (nA < rv[1])
   exp_1!(cache,A);
elseif (nA < rv[2]) 
   exp_2!(cache,A);
...
else 
   exp_15!(cache,A);
end
```
It's more code, but if you're saying the `eval` call is a problem it's okay with me.

Reply:
yeah, like that. Eval in Julia is dynamic and has to invoke the compiler, so there's a lot of issues there (it cannot optimize across the barrier, has a good amount of overhead, won't precompile, etc.) so it's best avoided in numerical codes. A big set of if statements is more code but less compute instructions.

Reply:
```julia
help?> Base.@nif
  @nif N conditionexpr expr
  @nif N conditionexpr expr elseexpr

  Generates a sequence of if ... elseif ... else ... end statements. For example:

  @nif 3 d->(i_d >= size(A,d)) d->(error("Dimension ", d, " too big")) d->println("All OK")

  would generate:

  if i_1 > size(A, 1)
      error("Dimension ", 1, " too big")
  elseif i_2 > size(A, 2)
      error("Dimension ", 2, " too big")
  else
      println("All OK")
  end
```
Can help you if you want elegant code.

Reply:
hmmm.. thanks. interesting. Elegant yes. Might be more difficult to read... at least for me.  :confused: